### PR TITLE
[RLlib] Exploration API ParameterNoise.

### PR DESCRIPTION
**Pull request by sven1977**
Move hard-coded parameter noise feature from DQN into new Exploration API.

The new methods ...

on_episode_start
on_episode_end
before_forward_pass
after_forward_pass
postprocess_trajectory

... have been added to enable the Exploration classes to control the ModelV2 before/after a forward pass (executed by the Policy) and to setup things before and after each episode.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: rllib/utils/exploration/parameter_noise.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+        else:
+            # Exploration: Normal behavior.
+            if explore:
+                return model(input_dict, state_batches, seq_lens)
+
+            # No exploration: Revert noise, forward-pass, then re-apply noise.
+            else:
+                self._remove_noise(model)
+                out = model(input_dict, state_batches, seq_lens)
+                self._apply_noise(model)
```
Comment (ericl) [inline rllib/utils/exploration/parameter_noise.py]: Is the only reason to override forward to wrap it in remove_noise / apply_noise?

If so, I think we can add less intrusive callbacks like `before_forward_pass` `after_forward_pass`.

Comment (ericl) [reply rllib/utils/exploration/parameter_noise.py]: (overall, I would prefer a small set of APIs that compose together rather than a monolithic override (unless we had no choice of course)).

Comment (ericl) [reply rllib/utils/exploration/parameter_noise.py]: (overall, I would prefer a smaller set of APIs that compose together rather than a monolithic override (unless we had no choice of course)).

Comment (sven1977) [reply rllib/utils/exploration/parameter_noise.py]: Yeah, maybe that's a less intrusive approach. I'll check whether this is possible. For param-noise, definitely, and it won't be much work to change this as it's only used in one Exploration class so far and the policy can just do again what's currently in `(base)Exploration.forward()`.

File: rllib/agents/dqn/dqn_policy.py [hunk 2]
```diff
@@ -362,12 +276,19 @@ def setup_late_mixins(policy, obs_space, action_space, config):
     TargetNetworkMixin.__init__(policy, obs_space, action_space, config)
 
 
-def _compute_q_values(policy, model, obs, obs_space, action_space):
+def _compute_q_values(policy, model, obs, explore):
     config = policy.config
+
+    policy.exploration.before_forward_pass(model, obs, explore=explore)
```
Comment (sven1977) [inline rllib/agents/dqn/dqn_policy.py]: This is an example for the before/forward/after split in a "manual" fashion (by providing a forward_fn in `build_tf_policy`).

Comment (sven1977) [reply rllib/agents/dqn/dqn_policy.py]: `forward_fn` replaces both `action_sampler_fn` AND `log_likelihood_fn`.

File: rllib/agents/dqn/tests/test_dqn.py [hunk 3]
```diff
@@ -99,6 +100,126 @@ def test_dqn_exploration_and_soft_q_config(self):
                 actions.append(trainer.compute_action(obs))
             check(np.std(actions), 0.0, false=True)
 
+    def test_dqn_parameter_noise_exploration(self):
```
Comment (sven1977) [inline rllib/agents/dqn/tests/test_dqn.py]: Added large test case here for ParameterNoise with DQN.

File: rllib/agents/sac/sac_policy.py [hunk 4]
```diff
@@ -95,37 +95,26 @@ def get_dist_class(config, action_space):
...
+        action_dist_class = (SquashedGaussian
+                             if config["normalize_actions"] else DiagGaussian)
     return action_dist_class
 
 
-def get_log_likelihood(policy, model, actions, input_dict, obs_space,
-                       action_space, config):
-    model_out, _ = model({
-        "obs": input_dict[SampleBatch.CUR_OBS],
+def get_distribution_inputs_and_class(policy,
```
Comment (sven1977) [inline rllib/agents/sac/sac_policy.py]: `forward_fn` (get_distribution_inputs_and_class) done for SAC as well. Replaces both `action_sampler_fn` and `log_likelihood_fn`.

File: rllib/evaluation/sampler.py [hunk 5]
```diff
@@ -306,6 +306,15 @@ def get_batch_builder():
     def new_episode():
         episode = MultiAgentEpisode(policies, policy_mapping_fn,
                                     get_batch_builder, extra_batch_callback)
+        # Call each policy's Exploration.on_episode_start method.
```
Comment (sven1977) [inline rllib/evaluation/sampler.py]: Sampler handles on_episode_start/end-calling via policy's Exploration.

File: rllib/policy/dynamic_tf_policy.py [hunk 6]
```diff
@@ -164,28 +160,48 @@ def __init__(self,
...
         else:
-            # Using an exploration setup.
-            sampled_action, sampled_action_logp = \
-                self.exploration.get_exploration_action(
-                    model_out,
-                    self.dist_class,
-                    self.model,
-                    timestep,
-                    explore=explore)
+            self.exploration.before_forward_pass(
```
Comment (sven1977) [inline rllib/policy/dynamic_tf_policy.py]: If no forward_fn given, perform default logic.

File: rllib/agents/ddpg/ddpg_policy.py [hunk 7]
```diff
@@ -74,6 +74,8 @@ def postprocess_trajectory(self,
 
 class DDPGTFPolicy(DDPGPostprocessing, TFPolicy):
     def __init__(self, observation_space, action_space, config):
+        self.observation_space = observation_space
```
Comment (sven1977) [inline rllib/agents/ddpg/ddpg_policy.py]: This will all be cleaned up properly and no longer necessary once we move DDPG into the functional Trainer API.

File: rllib/agents/dqn/distributional_q_model.py [hunk 8]
```diff
@@ -66,18 +71,17 @@ def build_action_value(model_out):
                     if use_noisy:
                         action_out = self._noisy_layer(
                             "hidden_%d" % i, action_out, q_hiddens[i], sigma0)
-                    elif parameter_noise:
+                    elif add_layer_norm:
```
Comment (sven1977) [inline rllib/agents/dqn/distributional_q_model.py]: These args (activation_fn and normalizer_fn) were outdated for tf2.x keras.

File: rllib/agents/dqn/dqn_policy.py [hunk 9]
```diff
@@ -124,34 +123,6 @@ def compute_td_error(obs_t, act_t, rew_t, obs_tp1, done_mask,
         self.compute_td_error = compute_td_error
 
 
-def postprocess_trajectory(policy,
```
Comment (sven1977) [inline rllib/agents/dqn/dqn_policy.py]: Moved into Exploration->ParameterNoise component.

File: rllib/agents/dqn/dqn.py [hunk 10]
```diff
@@ -221,49 +216,27 @@ def validate_config_and_setup_param_noise(config):
...
+            raise ValueError(
+                "ParameterNoise Exploration and `noisy` network cannot be "
+                "used at the same time!")
 
     # Update effective batch size to include n-step
     adjusted_batch_size = max(config["rollout_fragment_length"],
                               config.get("n_step", 1))
     config["rollout_fragment_length"] = adjusted_batch_size
 
-    # Setup parameter noise.
```
Comment (sven1977) [inline rllib/agents/dqn/dqn.py]: Moved into Exploration->ParameterNoise component.

File: rllib/agents/dqn/dqn_policy.py [hunk 11]
```diff
@@ -197,93 +172,44 @@ def build_q_model(policy, obs_space, action_space, config):
...
-def sample_action_from_q_network(policy, q_model, input_dict, obs_space,
-                                 action_space, explore, config, timestep):
-    # Action Q network.
-    q_vals = _compute_q_values(policy, q_model,
-                               input_dict[SampleBatch.CUR_OBS], obs_space,
-                               action_space)
-    policy.q_values = q_vals[0] if isinstance(q_vals, tuple) else q_vals
+    policy.q_values = q_vals
     policy.q_func_vars = q_model.variables()
-
```
Comment (sven1977) [inline rllib/agents/dqn/dqn_policy.py]: Moved into Exploration->ParameterNoise component.

File: rllib/agents/dqn/simple_q_policy.py [hunk 12]
```diff
@@ -22,15 +22,6 @@
 Q_TARGET_SCOPE = "target_q_func"
 
 
-class ParameterNoiseMixin:
```
Comment (sven1977) [inline rllib/agents/dqn/simple_q_policy.py]: Moved into Exploration->ParameterNoise component.

File: rllib/evaluation/sampler.py [hunk 13]
```diff
@@ -560,16 +576,24 @@ def _do_policy_eval(tf_sess, to_eval, policies, active_episodes):
                         TFPolicy.compute_actions.__code__):
             rnn_in_cols = _to_column_format(rnn_in)
             # TODO(ekl): how can we make info batch available to TF code?
-            # TODO(sven): Return dict from _build_compute_actions.
-            # it's becoming more and more unclear otherwise, what's where in
-            # the return tuple.
+            obs_batch = [t.obs for t in eval_data]
```
Comment (sven1977) [inline rllib/evaluation/sampler.py]: Sampler also handles before/after_forward_pass hooks for Exploration.

File: rllib/examples/rock_paper_scissors_multiagent.py [hunk 14]
```diff
@@ -202,11 +183,14 @@ def entropy_policy_gradient_loss(policy, model, dist_class, train_batch):
...
-    run_heuristic_vs_learned(use_lstm=True, trainer=EntropyLossPG)
+    run_heuristic_vs_learned(args, use_lstm=True, trainer=EntropyLossPG)
 
 
 if __name__ == "__main__":
-    # run_same_policy()
-    # run_heuristic_vs_learned(use_lstm=False)
-    run_heuristic_vs_learned(use_lstm=False)
-    # run_with_custom_entropy_loss()
+    parser = argparse.ArgumentParser()
```
Comment (sven1977) [inline rllib/examples/rock_paper_scissors_multiagent.py]: Added these calls, they were commented out for some reason.

File: rllib/policy/dynamic_tf_policy.py [hunk 15]
```diff
@@ -164,28 +163,32 @@ def __init__(self,
...
-        self.exploration = self._create_exploration(action_space, config)
         timestep = tf.placeholder(tf.int32, (), name="timestep")
 
-        # Setup custom action sampler.
-        if action_sampler_fn:
-            sampled_action, sampled_action_logp = action_sampler_fn(
-                self, self.model, self._input_dict, obs_space, action_space,
-                explore, config, timestep)
-        # Create a default action sampler.
+        if forward_fn:
```
Comment (sven1977) [inline rllib/policy/dynamic_tf_policy.py]: Custom forward_fn (only used in DQN and SAC) no longer need to deal with any exploration hooks!

File: rllib/tests/test_checkpoint_restore.py [hunk 16]
```diff
@@ -77,7 +77,7 @@ def ckpt_restore_test(use_object_store, alg_name, failures):
         alg2 = cls(config=CONFIGS[alg_name], env="CartPole-v0")
         env = gym.make("CartPole-v0")
 
-    for _ in range(2):
```
Comment (sven1977) [inline rllib/tests/test_checkpoint_restore.py]: Simplify this test; takes too much time otherwise.

File: rllib/policy/policy.py [hunk 17]
```diff
@@ -164,6 +166,38 @@ def compute_single_action(self,
         return action, [s[0] for s in state_out], \
             {k: v[0] for k, v in info.items()}
 
+    @DeveloperAPI
```
Comment (sven1977) [inline rllib/policy/policy.py]: Add this new method to Policy API will simplify and unify across different Algos:
1) log-likelihood calculations.
2) Action sampling.
3) Exploration handling.

**Review by ericl — CHANGES_REQUESTED** [review]
This is a pretty big PR with lots of changes. Logistically, can we break this down into a series of smaller PRs for review?
1. Add new param noise exploration interface.
2. PR to port DQN/SimpleQ.
3. PR to port remaining algos.
4. Other refactorings.

Design wise, I also think we can get away with a much more minimal interface for supporting exploration:
1. On episode start, on episode end hooks -> used to add and remove noise from the model.
2. Before compute action hook -> used to implement dynamic changing of the `exploration` argument in case the user needs to call compute_actions() outside an episode trajectory.

That would only require a single change to the Policy classes to call the before compute action hook, without introducing new abstractions or changing existing ones.

File: rllib/agents/a3c/a3c_torch_policy.py [hunk 18]
```diff
@@ -45,14 +45,12 @@ def add_advantages(policy,
...
-                              policy.config["use_gae"],
-                              policy.config["use_critic"])
+    return compute_advantages(
+        sample_batch, last_r, policy.config["gamma"], policy.config["lambda"],
+        policy.config["use_gae"], policy.config["use_critic"])
 
 
-def model_value_predictions(policy, input_dict, state_batches, model,
-                            action_dist):
+def model_value_predictions(policy, input_dict, state_batches, model):
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/a3c/a3c_torch_policy.py]: > This is a pretty big PR with lots of changes. Logistically, can we break this down into a series of smaller PRs for review?
> 
> 1. Add new param noise exploration interface.
ok, will split this into 2.

> 2. PR to port DQN/SimpleQ.
see above

> 3. PR to port remaining algos.
TODO later (DDPG)

> 4. Other refactorings.
> 
> Design wise, I also think we can get away with a much more minimal interface for supporting exploration:
> 
> 1. On episode start, on episode end hooks -> used to add and remove noise from the model.
> 2. Before compute action hook -> used to implement dynamic changing of the `exploration` argument in case the user needs to call compute_actions() outside an episode trajectory.

Yes, great idea, for parameter noise that would actually work, b/c we are already storing the noisy-state of the net and could only revert (before the next forward pass, if anything changed). We could get rid of after_forward_pass this way.

> That would only require a single change to the Policy classes to call the before compute action hook, without introducing new abstractions or changing existing ones.

**Comment by ericl** [comment]
Closing these for now since they build on #7772
